utils/models/idea-vault.js

```diff
@@ -0,0 +1,334 @@
+	return tiers.destroy({
+		where: {
+			channel: channel,
+		},
+	});
+};
+
+function insertIdea(msg, post) {
+	return ideas.create({
+		guild: msg.guild.id,
+		message: msg.id,
+		post: post.id,
+		post_channel: post.channel.id,
+	});
+};
+
+function upsertComment(id, author, value) {
+	return comments.upsert({
+		idea: id,
+		author: author,
+		value: value,
+	});
+};
+
+function getIdeaByMsg(msg) {
+	return ideas.findOne({
+		where: {
+			message: msg,
+		},
+	});
+};
+
+function getIdeaByPost(msg) {
+	return ideas.findOne({
+		where: {
+			post: msg,
+		},
+	});
+};
+
+function getIdeaByID(id) {
+	return ideas.findOne({
+		where: {
+			id: id,
+		},
+	});
+};
+
+function enable(guild) {
+	return guilds.upsert({
+		id: guild,
+		enabled: true,
+	});
+};
+
+function disable(guild) {
+	return guilds.upsert({
+		id: guild,
+		enabled: false,
+	});
+};
+
+function isEnabled(guild) {
+	return guilds.findOne({
+		attributes: ['enabled'],
+		where: {
+			id: guild,
+		},
+	}).then(result => {
+		return result.enabled;
+	});
+};
+
+async function generatePostEmbed(id, msg, count, comments = []) {
+	const embed = new MessageEmbed({
+		author: {
+			name: `${msg.author.username} in #${msg.channel.name}`,
+			icon_url: msg.author.avatarURL(),
+		},
+		description: msg.content,
+		footer: {
+			icon_url: 'https://emojipedia-us.s3.dualstack.us-west-1.amazonaws.com/thumbs/120/twitter/248/light-bulb_1f4a1.png',
+			text: `${count} | Idea #${id}`,
+		},
+		timestamp: msg.createdAt,
+	});
+
+	embed.addField('Original message', '[Here](' + msg.url + ')');
+
+	if (msg.attachments.size) {
+		const att = msg.attachments.first();
+		const imgtypes = ['jpg', 'jpeg', 'png', 'gif'];
+		if (imgtypes.includes(att.name.split('.').slice(-1)[0]).toLowerCase()) {
+			embed.setImage(att.url);
+		} else {
+			embed.addField('Attachments', att.url);
+		}
+	} else if (msg.embeds.length) {
+		const msgEmbed = msg.embeds[0];
+		switch (msgEmbed.type) {
+			case 'gifv':
+				embed.setImage(msgEmbed.url);
+				break;
+			case 'video':
+				embed.setTitle(msgEmbed.title);
+				break;
+
+			case 'link':
+				embed.setTitle(msgEmbed.title);
+				embed.setURL(msgEmbed.url);
+				embed.setThumbnail(msgEmbed.thumbnail.url);
+				break;
+
+			case 'rich':
+				if (msgEmbed.title) embed.setTitle(msgEmbed.title);
+				if (msgEmbed.description) embed.addField('Embed', msgEmbed.description);
+
+				embed.fields.push(...msgEmbed.fields);
+
+				if (msgEmbed.thumbnail) embed.setThumbnail(msgEmbed.thumbnail.url);
+				if (msgEmbed.image) embed.setImage(msgEmbed.image.url);
+				break;
+		}
+	}
+	for (const comment of comments) {
+		const author = await msg.guild.members.fetch(comment.author);
+		embed.addField('💬 Comment from ' + author.displayName, comment.value);
+	}
+
+	embed.setColor(msg.guild.me.displayColor || 16741829);
+	return embed;
+};
+
+async function messageReactionAdd(reaction, user) {
+	if (reaction.emoji.name !== '💡') return;
+
+	await reaction.fetch();
```

Comment:
Is reaction.emoji.name safe with partials?

And shouldn't reaction.fetch() be protected by a try/catch? In case a message got deleted?

Reply:
> Is reaction.emoji.name safe with partials?

Yes it is, if we look at the [official Discord documentation](https://discord.com/developers/docs/topics/gateway#message-reaction-add) we can see that the emoji is passed in the raw event. This means that in the worst case, where nothing was found in cache (partial) we can still expect `reaction.emoji.name`.
![image](https://user-images.githubusercontent.com/38372706/104137055-4bfef080-539a-11eb-9336-d7e78ccde369.png)


> And shouldn't reaction.fetch() be protected by a try/catch? In case a message got deleted?

Maybe? This should never happen. The only time it can happen is if the message was deleted right as someone light-bulbed it. I guess we can add a try/catch statement that'll just return if anything fails.
In that case there are a few more places where we could add try/catch statements. Though I am not sure, discord.js will usually handle it (like, the bot won't die because of that error).
This has not been done for the previous code:
https://github.com/enteryournamehere/wilson/blob/fffed159d6775d2c49ae01fe09a0d0039364eca4/winterbot.js#L143-L149

@enteryournamehere, do you think we should failsafe everything in that sense?

Reply:
Ah, that explains why the message event doesn't need a check for partial either.

One reason to might do a try/catch would be to prevent spamming the log with stuff that no one cares about. (Depending on how discord.js handles such exceptions)

Reply:
IMO it's fine to keep it like this. Even in the highly unlikely event that the message is deleted right after we receive this event, nothing bad will happen as it won't crash the entire bot. Probably not worth the time to add a try/catch to all places where something like this is used.